notifier/webhook/deliverer.go

```diff
@@ -0,0 +1,144 @@
+package webhook
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"path"
+	"time"
+
+	"github.com/google/uuid"
+	clairerror "github.com/quay/clair/v4/clair-error"
+	"github.com/quay/clair/v4/notifier"
+	"github.com/quay/clair/v4/notifier/keymanager"
+	"github.com/rs/zerolog"
+	"gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+type Deliverer struct {
+	conf Config
+	// a client to use for POSTing webhooks
+	c    *http.Client
+	kmgr *keymanager.Manager
+}
+
+// New returns a new webhook Deliverer
+func New(conf Config, client *http.Client, keymanager *keymanager.Manager) (*Deliverer, error) {
+	var c Config
+	var err error
+	if c, err = conf.Validate(); err != nil {
+		return nil, err
+	}
+	if client == nil {
+		client = http.DefaultClient
+	}
+	return &Deliverer{
+		conf: c,
+		c:    client,
+		kmgr: keymanager,
+	}, nil
+}
+
+func (d *Deliverer) Name() string {
+	return "webhook"
+}
+
+// sign will use the provided private key to sign and attach a jwt to the provided
+// request.
+//
+// a kid header specifying the private key which signed the jwt is attached as well.
+func (d *Deliverer) sign(ctx context.Context, req *http.Request, kp keymanager.KeyPair) error {
+	signer, err := jose.NewSigner(jose.SigningKey{Algorithm: jose.PS512, Key: kp.Private}, (&jose.SignerOptions{}).WithType("JWT"))
+	if err != nil {
+		return fmt.Errorf("failed to create jwt signer: %v", err)
+	}
+	now := jwt.NumericDate(time.Now().Unix())
+	expire := jwt.NumericDate(time.Now().Add(1 * time.Hour).Unix())
+	cl := jwt.Claims{
+		Subject:  "clair-notifier",
+		Issuer:   "clair-notifier",
+		Expiry:   &expire,
+		IssuedAt: &now,
+	}
+	kIDClaim := struct {
+		Kid string `json:"kid"`
+	}{
+		Kid: kp.ID.String(),
+	}
+	token, err := jwt.Signed(signer).Claims(cl).Claims(kIDClaim).FullSerialize()
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Authorization", "Bearer "+token)
+	req.Header.Set("kid", kp.ID.String())
```

Comment:
The keyserver protocol doesn't mandate this as an HTTP header.

Reply:
fixed